importers/src/main/groovy/whelk/importer/DatasetImporter.groovy

```diff
@@ -0,0 +1,135 @@
+        // succeeded (there may be intra-set dependencies). If the dataset contains circular dependencies,
+        // deletions will never be possible until the circle is unlinked/broken somewhere along the chain.
+        List<String> needsRetry = []
+        whelk.storage.doForIdInDataset(dataset, { String storedIdInDataset ->
+            if (!idsInInput.contains(storedIdInDataset)) {
+                if (!remove(whelk, storedIdInDataset)) {
+                    needsRetry.add(storedIdInDataset)
+                } else {
+                    deletedCount++
+                    if (deletedCount % 50 == 0) {
+                        System.err.println("Cleaning up: " + deletedCount + " records deleted (they are no longer in the dataset).")
+                    }
+                }
+            }
+        })
+
+        boolean anythingRemovedLastPass = true
+        while (anythingRemovedLastPass) {
+            anythingRemovedLastPass = false
+            needsRetry.removeAll { String storedIdInDataset ->
+                if (remove(whelk, storedIdInDataset)) {
+                    anythingRemovedLastPass = true
+                    deletedCount++
+                    return true
+                }
+                return false
+            }
+        }
+
+        if (!needsRetry.isEmpty()) {
+            log.warn("The following IDs SHOULD have been deleted, but doing so was not " +
+                    "possible, so they were skipped (most likely they are still depended upon):\n" + needsRetry)
+        }
+
+        System.err.println("Created: " + createdCount +" new,\n" +
+                "updated: " + updatedCount + " existing and\n" +
+                "deleted: " + deletedCount + " old records (should have been: " + (deletedCount + needsRetry.size()) + "),\n" +
+                "out of the: " + idsInInput.size() + " records in dataset: \"" + dataset + "\".\n" +
+                "Dataset now in sync.")
+    }
+
+    private static boolean remove(Whelk whelk, String id) {
+        try {
+            whelk.remove(id, "xl", null)
+            return true
+        } catch ( RuntimeException re ) {
+            // The expected exception here is: java.lang.RuntimeException: Deleting depended upon records is not allowed.
+            return false
+        }
+    }
+
+    private static void ensureAbsoluteSystemId(Document doc) {
+        def sysBaseIri = Document.BASE_URI
+        // A system id is in place; do nothing and return.
+        if (doc.recordIdentifiers.any { it.startsWith(sysBaseIri.toString()) }) {
+            return
+        }
+        // A relative system id (slug) exists; make it absolute and return.
+        def slug = doc.recordIdentifiers.find { it =~ /^\w+$/ }
+        if (slug) {
+            doc.setId(slug)
+            return
+        }
+        // Mint a new system Id and set it.
+        def newId = sysBaseIri.resolve(IdGenerator.generate()).toString()
+        doc.addRecordIdentifier(newId)
```

Comment:
Do we want to allow this case?
These will only be possible to import once (or import, read out, rebuild dataset I guess).

Reply:
Good catch! That code was copy pasted from the definitions importer, but you're right - the IDs can't be "random" like that! I'll fix it.